### DISCUSSION TITLE
Debug Adapter does not support stepping backwards with undodb

### DISCUSSION THREAD

**Discussion by SteveWolligandt** [mfussenegger/nvim-dap#893] [Q&A]
Hi!

With cpptools (OpenDebugAD7) and specifying `miDebuggerPath=udb` in the configuration nvim-dap says 
```
Debug Adapter does not support stepping backwards
```

I know that cpptools can do this since there is a plugin for VSCode for udb that uses cpptools.
(https://marketplace.visualstudio.com/items?itemName=Undo.udb)

I am using this config:
``` lua
function M.setup()
  local dap = require('dap')
  dap.adapters.cppdbg = {
    name = 'cppdbg',
    type = 'executable',
    command = vim.fn.stdpath('data') .. '/mason/bin/OpenDebugAD7',
  }
  dap.configurations.cpp = {
    {
      name = 'Debug Program',
      type = 'cppdbg',
      request = 'launch',
      MIMode = 'gdb',
      cwd = '${workspaceFolder}',
      udb='live',
      miDebuggerPath = 'udb',
      stopAtEntry = true,
      program = function()
        return M.inputAndWait('Select program')
      end,
    },
  }
end
```

Is that a bug of nvim-dap or is there really a flag that states that cpptools is not able to do reverse debugging?

Regards,
Steve

**Comment by SteveWolligandt** [comment]
When doing in the dap repl
```
-exec rn
```
for reverse next

I am getting this for like a thousand times:
```
Program received signal SIGSTOP, Stopped (signal).
The program has exited, but is still being debugged.
You can use UDB reverse commands to go backwards; see "help udb" for details.
0x00007ffff71b2da9 in _exit () from /lib64/libc.so.6
```

**Comment by SteveWolligandt** [comment]
That seems to be solvable in VSCode:
https://github.com/microsoft/vscode-cpptools/issues/2130#issuecomment-1147493724

Unfortunately it does not work for nvim-dap

**Comment by mfussenegger** [comment]
The debug adapter would have to set the `supportsStepBack` capability. Based on the linked issue cpptools currently doesn't set it.

But the linked workaround sounds like it could work.  What's the problem?

**Reply by SteveWolligandt** [reply]
Oh I just understood what that acutally does. Executing
``` lua
require'dap'.continue()
```
then actually steps back. At least this is working.
Do you maybe know how I get cpptools to set the `supportsStepBack` capability? It feels as if that beast is not really well documented.

**Reply by SteveWolligandt** [reply]
It seems the cpptools guys are planning to do that:
https://github.com/microsoft/MIEngine/pull/1037#issuecomment-881547051
However than I have another question. I saw that there is the possiblity to create custom commands in the repl. Unfortunately I could not figure out how to define a custom commands. Is there a way to do this?

Or is there a way to execute repl commands from lua? That would even be better!

 (BTW thanks for all the effort to develop nvim-dap. It is so useful!)

**Comment by SteveWolligandt** [comment]
By adding theses three lines in `lua/dap/repl.lua`
```lua
function M.exec(text)
  execute(text)
end
```
one can execute repl commands from lua by doing
```lua
local cmd = '...'
require('dap').repl.exec(cmd)
```
This would be a nice feature I think. I guess there are missing some sanity checks if a session is currently active.

**Comment by QuantumKuba** [comment]
Sorry for the bump, but was this added yet by any chance?
